### ISSUE TITLE
Ng2: Webpack + Maven build workflows

### ISSUE THREAD

**Issue by deepu105** [jhipster/generator-jhipster#4868]
A discussion about optimal dev workflow for the new Ng2 option started here https://github.com/jhipster/generator-jhipster/pull/4858#issuecomment-272567120

So I propose the below options. I would like to know if you guys agree @jhipster/developers @jdubois @gmarziou @VictorADS @cbornet 

**Generator:**
After generation is complete run `yarn install && yarn webpack:build` so that when maven/gradle dev build is run the webpack stuff is compiled and ready
Also output to the console asking users to start a `mvnw|gradlew` in a process and `yarn start` in another process for optimal development experience

**NPM commands:**

- `webpack:build` : compiles and builds all assets(vendor + main) for dev profile
- `webpack:dev `(alias start) : runs `webpack:build` and starts a watch/dev server with browsersync
- `start`: shorter alias for `webpack:dev`
- `webpack:prod` : runs lint, test and compiles & builds all assets(vendor + main) for prod profile

**Maven:**

- `dev`: runs only server dev build (similar to ng1 option, need to remove webpack build from this as it slows the deb build)
- `prod`: runs webpack prod build + server prod build

**Comment by jdubois** [comment]
There should be an option to run Webpack with only the "main" build: most of the time, the "vendor" will not move, so there is no need to rebuild it all the time, and that will make start up a lot faster.
Otherwise I totally agree.

**Comment by gmarziou** [comment]
Fully agree also.

> There should be an option to run Webpack with only the "main" build

Yes, it could be named "webpack:dev:main" or "webpack:dev:app"

angular-cli has 2 tests scripts, "test:watch" is really useful for TDD sessions.

~~~json
    "test": "karma start src/test/javascript/karma.conf.js",
    "test:watch": "karma start --watch",
~~~

**Comment by gmarziou** [comment]
> After generation is complete run `yarn install && yarn webpack:build` so that when maven/gradle dev build is run the webpack stuff is compiled and ready

Minor detail but could `yarn webpack:build` be run automatically by `postinstall` after install?

This way maybe we could have "webpack:dev" only watch at main as any change in deps would require a `yarn install`.

**Comment by deepu105** [comment]
pk we can an app only build option for people who wants that but not sure if many will use that as the time difference wont be that huge.

@gmarziou we already have `test` and `test:watch`
and yes the build ca be a postInstall step as well

**Comment by VictorADS** [comment]
Wouldn't it be easier if we could dynamically load the vendor with a shell script in the webpack config ?
That way running `yarn run webpack:dev` would create the vendor bundle only if it is not present

**Comment by deepu105** [comment]
Please no shell script magic

Thanks & regards,
Deepu

**Comment by VictorADS** [comment]
Even something as simple as this ? I'm only using node api 
```
const execSync = require('child_process').execSync;
const fs = require('fs');
const ddlPath = './target/www/vendor.json';
if (!fs.existsSync(ddlPath)) {
    execSync('webpack --config webpack/webpack.vendor.js');
}

**Comment by deepu105** [comment]
if you using only javascript and node API then no probs I thought you were
talking about bash scripts


Thanks & Regards,
Deepu

**Comment by jdubois** [comment]
@VictorADS is there anything else to do here? Let's close this !

**Comment by VictorADS** [comment]
@jdubois I guess you can close it

**Comment by deepu105** [comment]
@jdubois nope its not fully done yet as I still need to improve this as per my proposal. So plz keep it on

**Comment by jdubois** [comment]
OK @deepu105 keep us posted!

**Comment by Zacorich** [comment]
I had another problem that I could only solve by removing node_modules folder. I couldn't even get an error message as build process simply stuck first t webpack then on "yarn test && webpack -p --config webpack/webpack.vendor.js && webpack -p --config webpack/webpack.prod.js" so I tried to downgrade yarn and it worked but then stuck at "yarn run lint && karma start src/test/javascript/karma.conf.js" when I finally gave up and removed node_modules. I suggest to backup/archive node_modules directory just after first successful prod build and if there is any nodejs + yarn + webpack + java(maven) undocumented problem restore that folder from the archive instead of the internet any build again. For my setup it takes 15min to build prod from scratch with node_modules directory recreation so that's why I prefer archiving it.
